### PR TITLE
Support lets encrypt

### PR DESCRIPTION
**Pull request by alexcouper**
Requests to `well-known/acme-challenge/<challenge>` will be routed through to another container.

This allows you to have another container take care of all certificate gathering - for example this one:

https://github.com/ployst/docker/pull/2

### CODE REVIEW

File: start.sh [hunk 1]
```diff
@@ -37,8 +37,21 @@ if [ -n "${SERVICE_PORT_ENV_NAME+1}" ]; then
   TARGET_SERVICE="$TARGET_SERVICE:${!SERVICE_PORT_ENV_NAME}"
 fi
 
+# If the CERT_SERVICE_HOST_ENV_NAME and CERT_SERVICE_PORT_ENV_NAME vars
+# are provided, they point to the env vars set by Kubernetes that contain the
+# actual target address and port of the encryption service. Override the
+# default with them.
+if [ -n "${CERT_SERVICE_HOST_ENV_NAME+1}" ]; then
```
Comment (txels) [inline start.sh:44]: Why the `+1`?

File: nginx/proxy_ssl.conf [hunk 2]
```diff
@@ -2,14 +2,25 @@ upstream target_service {
   server {{TARGET_SERVICE}};
 }
 
+upstream cert_service {
+  server {{CERT_SERVICE}};
```
Comment (txels) [inline nginx/proxy_ssl.conf]: nice

File: start.sh (hunk 1 above)
Comment (alexcouper) [inline start.sh:44]: http://stackoverflow.com/a/13864829/1551613

File: nginx/proxy_ssl.conf (hunk 2 above)
Comment (txels) [inline nginx/proxy_ssl.conf]: Shouldn't this be an optional setting? Currently, if you don't pass in a `CERT_SERVICE` you can't start nginx:

```
$ docker run -e TARGET_SERVICE=localhost -e ENABLE_SSL=true 02c301a264d3
Enabling SSL...
Starting nginx...
2015/12/15 15:22:22 [emerg] 10#10: invalid number of arguments in "server" directive in /etc/nginx/conf.d/proxy.conf:6
nginx: [emerg] invalid number of arguments in "server" directive in /etc/nginx/conf.d/proxy.conf:6
```

It has to be a valid host as well:

```
$ docker run -e TARGET_SERVICE=10.0.2.2:7001 -e CERT_SERVICE=blah -e ENABLE_SSL=true 02c301a264d3
Enabling SSL...
Starting nginx...
2015/12/15 15:23:31 [emerg] 11#11: host not found in upstream "blah" in /etc/nginx/conf.d/proxy.conf:6
nginx: [emerg] host not found in upstream "blah" in /etc/nginx/conf.d/proxy.conf:6
```

Comment (txels) [inline nginx/proxy_ssl.conf]: I've tried a few things but there is no easy way to do this in plain nginx conf without using some template language/preprocessor to generate the configuration and omit sections if a variable is not defined...

Comment (alexcouper) [inline nginx/proxy_ssl.conf]: Hmm good point, I'll change this.

Comment (alexcouper) [inline nginx/proxy_ssl.conf]: Resolved in 5456960daea45c4ef26746a6762322a076e7d28a

**Pull request merged**